packages/ibm-products/src/components/CoachmarkV2/CoachmarkContent.tsx

```diff
@@ -0,0 +1,168 @@
+   */
+  className?: string;
+  /**
+   * This is a required callback that has to return the content to render in the body section.
+   * It can be a single child or an array of children depending on your need
+   */
+  children: ReactElement | ReactNode;
+  /**
+   * Specify whether the component should be rendered on high-contrast.
+   */
+  highContrast?: boolean;
+  /**
+   * Specify whether a drop shadow should be rendered on the popover.
+   */
+  dropShadow?: boolean;
+}
+
+export type CoachmarkContentComponent = React.ForwardRefExoticComponent<
+  CoachmarkContentProps & React.RefAttributes<HTMLDivElement>
+> & {
+  Header: React.FC<ContentHeaderProps>;
+  Body: React.FC<ContentBodyProps>;
+};
+
+const CoachmarkContent = React.forwardRef<
+  HTMLDivElement,
+  CoachmarkContentProps
+>((props, ref) => {
+  const { className = '', children, highContrast, dropShadow, ...rest } = props;
+  const coachmarkContentBlockClass = `${blockClass}--coachmark-content`;
+  const contentBodyClass = `${blockClass}--content-body`;
+  const { align, onClose, open, setOpen, triggerRef, setContentRef } =
+    useContext(CoachmarkV2Context);
+  const [targetId, setTargetId] = useState<string | null>(null);
+
+  // setting targetId from triggerRef context value
+  useLayoutEffect(() => {
+    if (open) {
+      const id = triggerRef?.current?.id ?? null;
+      setTargetId(id);
+    } else {
+      setTargetId(null);
+    }
+  }, [open, triggerRef]);
+
+  const bubbleRef = useRef<HTMLDivElement | null>(null);
+
+  useEffect(() => {
+    if (open && bubbleRef.current) {
+      requestAnimationFrame(() => {
+        const contentBody = bubbleRef.current!.querySelector(
+          `.${contentBodyClass}`
+        );
+
+        if (contentBody) {
+          const firstFocusable = Array.from(
+            contentBody.querySelectorAll<HTMLElement>('*')
+          ).find((el) => el.tabIndex >= 0);
+          firstFocusable?.focus();
+        }
+      });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [open]);
+
+  useEffect(() => {
+    const handleOutsideClick = (event: MouseEvent) => {
+      const targetElement = document.getElementById(targetId || '');
+      const bubbleElement = bubbleRef.current;
+
+      if (
+        bubbleElement &&
+        !bubbleElement.contains(event.target as Node) &&
+        targetElement &&
+        !targetElement.contains(event.target as Node)
+      ) {
+        setOpen?.(false);
+      }
+    };
+
+    if (open) {
+      document.addEventListener('click', handleOutsideClick);
+    }
+    return () => {
+      document.removeEventListener('click', handleOutsideClick);
+    };
+  }, [open, targetId, setOpen]);
+
+  useEffect(() => {
+    if (open && bubbleRef.current) {
+      const dragContainer = bubbleRef.current.querySelector(
+        `.${pkg.prefix}__bubble`
+      );
+      if (dragContainer instanceof HTMLElement) {
+        setContentRef(dragContainer);
```

Comment:
i think i mentioned this elsewhere, but we should probably just be setting this ref with `useRef` instead of `useState`